File: 1054-Distant Barcodes.py
class Solution(object):
    def rearrangeBarcodes(self, barcodes):
        """
        :type barcodes: List[int]
        :rtype: List[int]
        """

# O(nlogn)
        N = len(barcodes)
        res = [0]*N
        ind = 0
        
        count = collections.Counter(barcodes)
        for code in sorted(count.keys(), key = lambda x: -count[x]):
            for _ in range(count[code]):
                res[ind] = code
                ind += 2
                if ind >= N:
                    ind = 1
                    
        return res

File: JavaScript/1054-Distant Barcodes.js
/**
 * @param {number[]} barcodes
 * @return {number[]}
 */
var rearrangeBarcodes = function(barcodes) {
    const N = barcodes.length;
    let res = [];
    let hmap = new Map();
    
    for (let code of barcodes){
        hmap.set(code, (hmap.get(code)||0) + 1);
    }
    
    let arr = Array.from(hmap.keys()).sort(function(a, b){
        return hmap.get(b) - hmap.get(a);
    })
    
    let index = 0;
    
    for (let key of arr){
        for (let i=0; i<hmap.get(key); i++){
            res[index] = key;
            index += 2;
            if (index>=N){
                index = 1;
            }
        }
    }
    
    return res;
};